src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift

```diff
@@ -0,0 +1,303 @@
+                     maxInputLength: UInt(kMaxNumOfVariableNameCharacters),
+                     isList: false,
+                     andTextField: nil,
+                     initialText: "")
+
+    }
+
+    func askForNewVariableName() {
+        Util.askUser(forVariableNameAndPerformAction: #selector(saveVariable(name:)),
+                     target: self,
+                     promptTitle: kUIFENewVarExists,
+                     promptMessage: kUIFEOtherName,
+                     minInputLength: UInt(kMinNumOfVariableNameCharacters),
+                     maxInputLength: UInt(kMaxNumOfVariableNameCharacters),
+                     isList: false,
+                     andTextField: nil,
+                     initialText: "")
+    }
+
+    @objc func saveVariable(name: String) {
+        if self.newVarIsForProject {
+            if let project = self.spriteObject?.scene.project {
+                for variable in UserDataContainer.allVariables(for: project) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        } else {
+            if let object = self.spriteObject {
+                for variable in UserDataContainer.objectAndProjectVariables(for: object) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        }
+
+        let userVariable = UserVariable(name: name)
+        userVariable.value = Int(0)
+
+        if self.newVarIsForProject {
+            self.spriteObject?.scene.project?.userData.add(userVariable)
+        } else {
+            self.spriteObject?.userData.add(userVariable)
+        }
+
+        self.spriteObject?.scene.project?.saveToDisk(withNotification: false)
+        self.reloadData()
+    }
+
+    @objc func saveList(name: String) {
+        if self.newVarIsForProject {
+            if let project = self.spriteObject?.scene.project {
+                for variable in UserDataContainer.allLists(for: project) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        } else {
+            if let object = self.spriteObject {
+                for variable in UserDataContainer.objectAndProjectLists(for: object) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        }
+
+        let userList = UserList(name: name)
+
+        if self.newVarIsForProject {
+            self.spriteObject?.scene.project?.userData.add(userList)
+        } else {
+            self.spriteObject?.userData.add(userList)
+        }
+
+        self.spriteObject?.scene.project?.saveToDisk(withNotification: false)
+        self.reloadData()
+    }
+
+    func isVariableUsed(_ userVariable: UserVariable) -> Bool {
+
+        guard let project = self.spriteObject?.scene.project else {
```

Comment:
Could you please come up with a unit test which is verifying the behavior of `isVariableUsed` and `isListUsed`? Just make sure that this behavior is not tested using a UI test. Once we have a regression test, I am sure we can do some refactoring here.

Reply:
We already have `FormulaEditorSectionViewControllerTests` where `isVariableUsed` and `isListUsed` is tested. Is that what you are asking or am I misunderstanding?

Reply:
Perfect! So let's start refactoring: I would suggest to get rid of a lot of duplicated code by introducing the following two methods:

`- private func getAllBricks(for object: SpriteObject) -> [Brick]`

This method could be used for checking variables/lists with both global and local scope.